Sources/APIProvider.swift

```diff
@@ -51,111 +57,85 @@ public final class APIProvider {
-    @discardableResult
-    public func request<T: Decodable>(_ endpoint: APIEndpoint<T>, completion: @escaping RequestCompletionHandler<T>) -> DataRequest {
-        let dataRequest = defaultSessionManager.request(endpoint)
-        dataRequest.mapResponseTo(T.self, decoder: jsonDecoder, completion: completion).resume()
-        return dataRequest
+    public func request<T: Decodable>(_ endpoint: APIEndpoint<T>, completion: @escaping RequestCompletionHandler<T>) {
+        guard let request = try? self.requestsAuthenticator.adapt(endpoint.asURLRequest()) else {
+            completion(.failure(Error.requestGeneration))
+            return
+        }
+
+        self.requestExecutor.execute(request) { completion(self.mapResponse($0)) }
     }
     
     /// Performs a data request to the given ResourceLinks
     ///
     /// - Parameters:
     ///   - resourceLinks: The resourceLinks to request.
     ///   - completion: The completion callback which will be called on completion containing the result.
-    @discardableResult
-    public func request<T: Decodable>(_ resourceLinks: ResourceLinks<T>, completion: @escaping RequestCompletionHandler<T>) -> DataRequest {
-        let dataRequest = defaultSessionManager.request(resourceLinks.`self`)
-        dataRequest.mapResponseTo(T.self, decoder: jsonDecoder, completion: completion).resume()
-        return dataRequest
+    public func request<T: Decodable>(_ resourceLinks: ResourceLinks<T>, completion: @escaping RequestCompletionHandler<T>) {
+
+        self.requestExecutor.retrieve(resourceLinks.`self`) { completion(self.mapResponse($0)) }
     }
 }
 
-extension DataRequest {
-    
-    /// Defines errors which are caused on JSON mapping.
-    enum JSONMappingError: Error {
-        /// Indicates that the response is not a valid JSON dictionary.
-        case invalidResponse
-    }
-    
-    /// Maps the response to the given JSONDecodable data type.
-    ///
-    /// - Parameters:
-    ///   - type: The type to map to.
-    ///   - completion: The result of the mapping. An error will be returned if mapping fails.
-    @discardableResult
-    func dataResponse(decoder: JSONDecoder, completion: @escaping RequestCompletionHandler<Data?>) -> Self {
-        return validate(statusCode: 200..<300).responseData(queue: DispatchQueue.global(qos: .background)) { response in
-            if let error = response.error {
-                // Try to parse api error
-                guard let data = response.data, let apiError = try? decoder.decode(ErrorResponse.self, from: data) else {
-                    completion(Result.failure(error))
-                    return
-                }
-                completion(Result.failure(apiError))
-            } else {
-                completion(Result.success(response.data))
+// MARK: - Private
+
+private extension APIProvider {
+
+    func mapResponse<T: Decodable>(_ result: Result<Response>) -> Result<T> {
```

Comment:
Could you add the documentation to these methods as well, just like it was before?